### PR TITLE
Add MACE model support

### PR DESCRIPTION
**Pull request by sef43**
This adds support for MACE models: https://github.com/ACEsuit/mace.

The MACE model framework is well defined. With this PR a user can run an OpenMM simulation using a MACE model trained with the scripts in https://github.com/ACEsuit/mace.

cc @jharrymoore @jmichel80

### CODE REVIEW

File: openmmml/models/macepotential.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+                  forceGroup: int,
+                  implementation : str=None,
+                  dtype: str=None,
+                  interaction_energy: bool=True,
+                  **args):
+        
+
+        import torch
+        import openmmtorch
+        from e3nn.util import jit
```
Comment (raimis) [inline openmmml/models/macepotential.py:84]: This should give a better error message than import failed. An explanation of missing MACE and pip command should be shown.

File: openmmml/models/macepotential.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        if atoms is not None:
+            includedAtoms = [includedAtoms[i] for i in atoms]
+        
+        class MACEForce(torch.nn.Module):
+
+            def __init__(self, model_path, atomic_numbers, indices, periodic, dtype=None, interaction_energy=True):
+                super(MACEForce, self).__init__()
+
+
+                if not torch.cuda.is_available():
```
Comment (raimis) [inline openmmml/models/macepotential.py:97]: OpenMM-Torch takes care of loading to the right devices. Here you just need to load on CPU.

File: openmmml/models/macepotential.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                    self.model = torch.load(model_path, map_location=torch.device('cpu'))
+                else:
+                    self.model = torch.load(model_path)
+                
+                # get the dtype of the saved model
+                model_dtype = [p.dtype for p in self.model.parameters()][0]
+
+                # check is user has request a different dtype
```
Comment (raimis) [inline openmmml/models/macepotential.py:105]: The model should run on the precision with which it was trained. If OpenMM runs at a different precision (double rather float), one just casts inputs/outputs accordingly rather than change the model.

File: openmmml/models/macepotential.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+                # check is user has request a different dtype    
+                if dtype is not None and dtype != model_dtype:
+                    logging.getLogger().warning(f'Warning: Loaded MACE model has dtype of {model_dtype} which is being changed to {dtype} as requested!')
+                    self.default_dtype = dtype
+                    self.model.to(self.default_dtype)
+                else:
+                    self.default_dtype = model_dtype
+
+                torch.set_default_dtype(self.default_dtype)
```
Comment (raimis) [inline openmmml/models/macepotential.py:113]: This a global PyTorch setting. It is better to set tensor precisions explicitly.

File: openmmml/models/macepotential.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+
+                #self.register_buffer('r_max', self.model.r_max)
+                self.z_table = utils.AtomicNumberTable([int(z) for z in self.model.atomic_numbers])
+
+                self.model = jit.compile(self.model)
+                
+                # setup input
+                N = len(atomic_numbers)
+                self.ptr = torch.nn.Parameter(torch.tensor([0,N], dtype=torch.long), requires_grad=False)
```
Comment (raimis) [inline openmmml/models/macepotential.py:128]: Why these are parameters rather than buffers?

File: openmmml/models/macepotential.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        import torch
+        import openmmtorch
+        from e3nn.util import jit
+        from mace.tools import utils, to_one_hot, atomic_numbers_to_indices
+        
+        includedAtoms = list(topology.atoms())
+        if atoms is not None:
+            includedAtoms = [includedAtoms[i] for i in atoms]
+        
+        class MACEForce(torch.nn.Module):
```
Comment (raimis) [inline openmmml/models/macepotential.py:91]: For better readability, don't nest the classes.

File: openmmml/models/macepotential.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+                              "batch": self.batch, 
+                              "pbc": self.pbc,
+                              "cell": cell,
+                              "positions": positions,
+                              "edge_index": edge_index,
+                              "unit_shifts": shifts_idx,
+                              "shifts": shifts}
+                
+                # predict
+                out = self.model(input_dict,compute_force=False)
```
Comment (raimis) [inline openmmml/models/macepotential.py:187]: Why do the forces are computed here?

File: openmmml/models/macepotential.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+                energy = out[self.return_energy_type]
+                if energy is None:
```
Comment (raimis) [inline openmmml/models/macepotential.py:190]: How can you get energy `None`? The simulation should fail rather than silently assume it is 0.

File: openmmml/models/utils.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
+import torch
+from typing import Tuple
+
+@torch.jit.script
+def simple_nl(positions: torch.Tensor, cell: torch.Tensor, pbc: bool, cutoff: float, sorti: bool=False) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (raimis) [inline openmmml/models/utils.py:5]: Do we really need to invent a new neighbor list rather than use the one in NNPOps?

**Review by raimis — CHANGES_REQUESTED** [review]

**Comment by JMorado** [comment]
I have been working on this PR and would like to discuss the best approach for incorporating the computation of the neighbour list. There are 3 possible options:

1. Utilising the `getNeighborPairs` function from the `NNPOps` module.
2. Utilising the `simple_nl` function introduced in the `utils` module, which is included with this PR.
3. Include both and leave the choice to the user.

What do you think would be preferable? The first approach has the advantage of not needing additional modules to be added to `openmm-ml`, but it does make `NNPOps` a mandatory dependency of `openmm-ml`. Additionally, it necessitates recalculating the vectors between neighbors during the forward pass for shift computation. Perhaps integrating the computation of shifts into the `NNPOps` implementations could be beneficial, as it may also prove useful for other MLPs.

**Comment by peastman** [comment]
> but it does make NNPOps a mandatory dependency of openmm-ml.

It's not mandatory: it's only needed if you use MACE.  I think that's ok.  I expect this package to eventually have interfaces to a lot of other libraries, each of which might or might not be installed.  So for example if MACE is installed, you can create MACE models.  And if it isn't installed, that doesn't stop you from creating other types of models.

> Additionally, it necessitates recalculating the vectors between neighbors during the forward pass for shift computation.

Compared to the cost of computing the model, computing the vector between two atoms should be totally negligible!

**Comment by JMorado** [comment]
Sounds good. And I agree that the calculation of the vector should be negligible.

**Comment by JMorado** [comment]
This is ready for review. I've addressed most of the comments raised by @raimis. In addition to general refactoring, I have removed the `utils.py` module as now `getNeighborPairs` from `NNPops` is used.

**Review by peastman — COMMENTED** [review]
This looks great!  Just a few very minor comments.

**Comment by JMorado** [comment]
Thank you for the review, @peastman. All fixed now.

**Comment by JMorado** [comment]
I've also added a fix to ensure tensors are placed on the correct devices. I was only testing the implementation on the CPU and Reference platforms and was not having any issues. I've now conducted testing across all platforms, and everything should be functioning correctly.

File: openmmml/models/macepotential.py [hunk 10]
```diff
@@ -52,13 +52,13 @@ class MACEPotentialImpl(MLPotentialImpl):
...
 
-    Other available MACE-OFF23 models include `mace-off23-medium` and `mace-off23-large`.
+    Other available MACE-OFF23 models include 'mace-off23-medium' and 'mace-off23-large'.
 
     To use a locally trained MACE model, provide the path to the model file. For example:
     >>> potential = MLPotential('mace', modelPath='MACE.model')
 
     During system creation, you can optionally specify the precision of the model using the
-    `precision` keyword argument. Supported options are 'single' and 'double'. For example:
+    precision` keyword argument. Supported options are 'single' and 'double'. For example:
```
Comment (peastman) [inline openmmml/models/macepotential.py:61]: You accidentally removed the quote at the start of the line.

File: openmmml/models/macepotential.py [hunk 11]
```diff
@@ -0,0 +1,413 @@
...
+            dtype,
+            returnEnergyType,
+        )
+
+        # Convert it to TorchScript and save it.
+        module = torch.jit.script(maceForce)
+        module.save(filename)
+
+        # Create the TorchForce and add it to the System.
+        force = openmmtorch.TorchForce(filename)
```
[the remark this answers is not in the document]
Comment (peastman) [reply openmmml/models/macepotential.py]: Yes, it would probably be good to clean that up eventually.

File: openmmml/models/macepotential.py (hunk 10 above)
Comment (JMorado) [reply openmmml/models/macepotential.py:61]: Ups! Fixed.

File: openmmml/models/macepotential.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (JMorado) [reply openmmml/models/macepotential.py]: Ok, I'll take care of it.

**Comment by peastman** [comment]
Looks good.  Is this ready to merge?

**Comment by JMorado** [comment]
Ready to merge.

**Comment by peastman** [comment]
Thanks!
